VitruvioHost/Plugins/Vitruvio/Source/Vitruvio/Private/Util/PolygonWindings.cpp

```diff
@@ -17,99 +17,254 @@
+	TSortedMap<int32, FWindingEdge> WindingEdgeMap;
+	for (auto& KeyValue : EdgeMap)
 	{
-		if (Edge.Count == 1)
+		for (const FWindingEdge& Current : KeyValue.Value)
 		{
-			EdgeMap.Add(Edge.Index0, Edge);
+			const FWindingEdge Opposite(Current.Index1, Current.Index0, Current.Connectivity);
+			if (!Edges.Contains(Opposite))
+			{
+				// Note that at this point there should not be multiple edges connected to a single vertex
+				WindingEdgeMap.Add(Current.Index0, Current);
+			}
 		}
 	}
 
-	// Organize the remaining edges in the list so that the vertices will meet up to form a continuous outline around the shapes
-	TArray<TArray<int32>> Windings;
-	while (EdgeMap.Num() > 0)
+	// Organize the remaining edges in the list so that the vertices will meet up to form a continuous outline of either a face or a hole
+	TArray<FWinding> Windings;
+	while (WindingEdgeMap.Num() > 0)
 	{
 		TArray<int32> WindingIndices;
 
 		// Get and remove first edge
-		auto EdgeIter = EdgeMap.CreateIterator();
+		auto EdgeIter = WindingEdgeMap.CreateIterator();
 		const FWindingEdge FirstEdge = EdgeIter.Value();
 		EdgeIter.RemoveCurrent();
 
 		WindingIndices.Add(FirstEdge.Index0);
 		int NextIndex = FirstEdge.Index1;
 
 		// Find connected edges
-		while (EdgeMap.Contains(NextIndex))
+		while (WindingEdgeMap.Contains(NextIndex))
 		{
-			const FWindingEdge& Current = EdgeMap.FindAndRemoveChecked(NextIndex);
+			const FWindingEdge& Current = WindingEdgeMap.FindAndRemoveChecked(NextIndex);
 			WindingIndices.Add(Current.Index0);
 			NextIndex = Current.Index1;
 		}
 
-		Windings.Add(WindingIndices);
+		Windings.Add({WindingIndices, FirstEdge.Connectivity});
+	}
+
+	// Find the relation between the faces
+	TMap<int32, int32> InsideOf;
+	for (int32 IndexA = 0; IndexA < Windings.Num(); IndexA++)
+	{
+		for (int32 IndexB = IndexA + 1; IndexB < Windings.Num(); IndexB++)
+		{
+			// No possible relation if they are not connected
+			if (Windings[IndexA].Connectivity != Windings[IndexB].Connectivity)
+			{
+				continue;
+			}
+
+			if (IsInsideOf2D(Windings[IndexA].Indices, Windings[IndexB].Indices, InVertices))
+			{
+				InsideOf.Add(IndexA, IndexB);
+			}
+			else if (IsInsideOf2D(Windings[IndexB].Indices, Windings[IndexA].Indices, InVertices))
+			{
+				InsideOf.Add(IndexB, IndexA);
+			}
+		}
+	}
+
+	TSet<int32> OpenSet;
+	for (int32 FaceIndex = 0; FaceIndex < Windings.Num(); ++FaceIndex)
+	{
+		OpenSet.Add(FaceIndex);
+	}
+
+	TMap<int32, FFace> FaceMap;
+	// First find all faces
+	for (int32 FaceIndex = 0; FaceIndex < Windings.Num(); ++FaceIndex)
+	{
+		if (!InsideOf.Contains(FaceIndex))
+		{
+			FaceMap.Add(FaceIndex, FFace{Windings[FaceIndex].Indices});
+			OpenSet.Remove(FaceIndex);
+		}
+	}
+
+	// Then assign the holes
+	for (int32 FaceIndex : OpenSet)
+	{
+		int32 InsideOfIndex = InsideOf[FaceIndex];
+		if (FaceMap.Contains(InsideOfIndex))
+		{
+			FFace& ParentFace = FaceMap[InsideOfIndex];
+			ParentFace.Holes.Add(FHole{Windings[FaceIndex].Indices});
+		}
```

Comment:
The above 3 for loops could technically be embedded in the upper for loop (since after iterating over all A-B matches (inner for loop) we know if A contains holes (and which ones) or is a hole itself). 
I don't mind if we keep them outside if you think it improves readability, but it makes the GetPolygon function more verbose than necessary.

Reply:
Simplified that as well. Not to one loop but two. Should be more readable now.